File: listas.java
class Nodo {
    int dato;
    Nodo sig;

    Nodo(int v) {
        dato = v;
        sig = null;
    }
}

public class Lista {
    Nodo inicio;

    public Lista() {
        inicio = null;
    }

    public void insertarInicio(int v) {
        Nodo nuevo = new Nodo(v);
        nuevo.sig = inicio;
        inicio = nuevo;
    }

    public void insertarFinal(int v) {
        Nodo nuevo = new Nodo(v);
        if (inicio == null) {
            inicio = nuevo;
            return;
        }
        Nodo aux = inicio;
        while (aux.sig != null) aux = aux.sig;
        aux.sig = nuevo;
    }

    public boolean buscar(int v) {
        Nodo aux = inicio;
        while (aux != null) {
            if (aux.dato == v) return true;
            aux = aux.sig;
        }
        return false;
    }

    public void eliminar(int v) {
        if (inicio == null) return;

        if (inicio.dato == v) {
            inicio = inicio.sig;
            return;
        }

        Nodo aux = inicio;
        while (aux.sig != null && aux.sig.dato != v)
            aux = aux.sig;

        if (aux.sig != null)
            aux.sig = aux.sig.sig;
    }

    public void mostrar() {
        Nodo aux = inicio;
        while (aux != null) {
            System.out.print(aux.dato + " -> ");
            aux = aux.sig;
        }
        System.out.println("NULL");
    }

    public static void main(String[] args) {
        Lista l = new Lista();
        l.insertarInicio(10);
        l.insertarFinal(20);
        l.insertarFinal(30);
        l.mostrar();
    }
}

File: listas.py
class Nodo:
    def __init__(self, v):
        self.dato = v
        self.sig = None

class Lista:
    def __init__(self):
        self.inicio = None

    def insertar_inicio(self, v):
        nuevo = Nodo(v)
        nuevo.sig = self.inicio
        self.inicio = nuevo

    def insertar_final(self, v):
        nuevo = Nodo(v)
        if self.inicio is None:
            self.inicio = nuevo
            return
        aux = self.inicio
        while aux.sig:
            aux = aux.sig
        aux.sig = nuevo

    def buscar(self, v):
        aux = self.inicio
        while aux:
            if aux.dato == v:
                return True
            aux = aux.sig
        return False

    def eliminar(self, v):
        if self.inicio is None:
            return

        if self.inicio.dato == v:
            self.inicio = self.inicio.sig
            return

        aux = self.inicio
        while aux.sig and aux.sig.dato != v:
            aux = aux.sig

        if aux.sig:
            aux.sig = aux.sig.sig

    def mostrar(self):
        aux = self.inicio
        while aux:
            print(aux.dato, "->", end=" ")
            aux = aux.sig
        print("NULL")

# Ejemplo
l = Lista()
l.insertar_inicio(10)
l.insertar_final(20)
l.insertar_final(30)
l.mostrar()
